### PR TITLE
Story 751

### LINKED ISSUE #751: As a GEOPM developer, I would like a C++ helper class for parsing command line options
Use Case
----------
The motivation for this is to avoid having to write command line tools in C, to hide the getopt logic behind a testable class, and to support refactoring commandline tools with too much untested logic in main().

Functionality should be similar to Python's argparse.

Explicitly not supported:
- named positional arguments
- long form options with single hyphen
- optional arguments to options (see #749)
- required options

Done Criteria
---------------
- New OptionParser class with unit tests.

### PR DESCRIPTION
**Pull request by dianarg**
- Resolves #751

### CODE REVIEW

File: src/OptionParser.hpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    class OptionParser
+    {
+        public:
+            OptionParser(const std::string &prog_name,
+                         std::ostream &std_out,
+                         std::ostream &err_out,
+                         const std::string &custom_help);
+            /// @brief Add an option.  Boolean types do not require an
+            ///        argument; all other types do.
+            template <typename T>
```
Comment (dianarg) [inline src/OptionParser.hpp]: No template

File: src/OptionParser.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+#include <iostream>
+namespace geopm
+{
+    static const std::string COPYRIGHT_TEXT = "\nCopyright (c) 2015, 2016, 2017, 2018, 2019, Intel Corporation. All rights reserved.\n\n";
+
+    OptionParser::OptionParser(const std::string &prog_name,
+                               std::ostream &std_out,
+                               std::ostream &err_out,
+                               const std::string &custom_help="")
```
Comment (dianarg) [inline src/OptionParser.cpp]: another constructor

File: src/OptionParser.cpp [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        : m_prog_name(prog_name)
+        , m_std_out(std_out)
+        , m_err_out(err_out)
+        , m_custom_help(custom_help)
+    {
+        // automatically support --help and --version
+        add_option("help", 'h', "help", false,
+                   "print brief summary of the command line usage information, then exit");
+        add_option("version", 'v', "version", false,
```
Comment (dianarg) [inline src/OptionParser.cpp]: V instead of v

File: src/OptionParser.cpp [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        bool do_help = false;
+        bool do_version = false;
+        std::ostringstream msg;
+        int opt;
+        optind = 1;  // reset to allow multiple parse() calls
+        while (!err && (opt = getopt_long(argc, (char * const *)argv, short_options.c_str(), long_options.data(), NULL)) != -1) {
+            if (opt == 'h') {
+                do_help = true;
+            }
+            else if (opt == 'v') {
```
Comment (dianarg) [inline src/OptionParser.cpp]: 'V'

File: src/OptionParser.cpp [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+#include <sstream>
+#include <algorithm>
+
+#include "geopm_version.h"
+#include "config.h"
+
+#include <iostream>
+namespace geopm
+{
+    static const std::string COPYRIGHT_TEXT = "\nCopyright (c) 2015, 2016, 2017, 2018, 2019, Intel Corporation. All rights reserved.\n\n";
```
Comment (dianarg) [inline src/OptionParser.cpp]: G_COPYRIGHT_TEXT

Comment (dianarg) [reply src/OptionParser.cpp]: could also be static member data of class: M_COPYRIGHT_TEXT

File: src/OptionParser.cpp [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        tmp << usage_start << " [--help] [--version]\n";
+        tmp << "\nMandatory arguments to long options are mandatory for short options too.\n\n";
+
+        // The first column of options starts indented to col0.  The
+        // second column of descriptions starts at col1 and wraps if
+        // longer than col2.  If the first column is wider than col1,
+        // the description will start on the next line.
+        const int col0 = 2;
+        const int col1 = 28;
+        const int col2 = 80;
```
Comment (dianarg) [inline src/OptionParser.cpp]: 80-1

File: src/OptionParser.cpp [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+            std::string right = "";
+            while (description.size() > col2 - col1) {
+                auto end = description.find_last_of(" ", col2-col1);
+                auto sub = description.substr(0, end);
+                tmp << left << sub << "\n";
+                left = std::string(col1, ' ');
+                description = description.substr(end+1, std::string::npos);
+            }
+            tmp << left << description << "\n";
+        }
```
Comment (dianarg) [inline src/OptionParser.cpp]: check for binary operators without spaces

Comment (dianarg) [reply src/OptionParser.cpp]: this could be a helper, move col variables in here

**Review by bakerbrandond — APPROVED** [review]

**Pull request merged**